runtime/calamari/src/zenlink.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2020-2023 Manta Network.
+// This file is part of Manta.
+//
+// Manta is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::{AssetManager, Balances, ParachainInfo, Runtime, RuntimeEvent, ZenlinkProtocol};
+use crate::assets_config::CalamariConcreteFungibleLedger;
+use frame_support::{parameter_types, traits::ExistenceRequirement, PalletId};
+use manta_primitives::{
+    assets::{AssetIdLocationMap, AssetIdLpMap, FungibleLedger},
+    types::CalamariAssetId,
+};
+use sp_runtime::DispatchError;
+use zenlink_protocol::{
+    AssetBalance, AssetId as ZenlinkAssetId, GenerateLpAssetId, LocalAssetHandler,
+    ZenlinkMultiAssets, LOCAL,
+};
+
+// Normal Coin AMM
+parameter_types! {
+    pub const ZenlinkPalletId: PalletId = PalletId(*b"/zenlink");
+    pub SelfParaId: u32 = ParachainInfo::parachain_id().into();
+    pub MantaNativeAssetId: CalamariAssetId = 1;
+    pub ZenlinkNativeAssetId: u64 = 0;
+}
+
+type MultiAssets = ZenlinkMultiAssets<ZenlinkProtocol, Balances, LocalAssetAdaptor>;
+
+impl zenlink_protocol::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type MultiAssetsHandler = MultiAssets;
+    type PalletId = ZenlinkPalletId;
+    type SelfParaId = SelfParaId;
+    type AssetId = ZenlinkAssetId;
+    #[cfg(not(feature = "runtime-benchmarks"))]
+    type LpGenerate = AssetManagerLpGenerate;
+    #[cfg(feature = "runtime-benchmarks")]
+    type LpGenerate = mock_benchmark::MockAssetManagerLpGenerate;
+    type WeightInfo = ();
+}
+
+pub struct AssetManagerLpGenerate;
+impl GenerateLpAssetId<ZenlinkAssetId> for AssetManagerLpGenerate {
+    fn generate_lp_asset_id(
+        asset_0: ZenlinkAssetId,
+        asset_1: ZenlinkAssetId,
+    ) -> Option<ZenlinkAssetId> {
+        // LP asset id is registered on AssetManager based on two asset id.
+        let asset_id_0 = LocalAssetAdaptor::asset_id_convert(asset_0);
```

Comment:
I'm not sure it will happen.
What if `asset_0` == `asset_1`, and asset_0 is valid?

Reply:
Add asset0 != asset1 check condition. in theory, this could't happen.